File: js/example/remote-engine.mjs
// bun run --preload ../../../engine-js/bun-loader.mjs ./remote-engine.mjs
// node --enable-source-maps --import ../../../engine-js/nodejs-loader.js ./remote-engine.mjs

// Comment this if you want to use published version of the module
import '../../../dist/engine.min.js';
import '../../../dist/remote-control.min.js';

// import '@blackprint/engine';
// import '@blackprint/remote-control';

import { createServer } from "http";
import { Server } from "socket.io";

globalThis.window = globalThis;

let port = 2345;
let httpServer = createServer();
let io = new Server(httpServer, {
  cors: {
    origin: ["http://localhost:6789", "https://blackprint.github.io"],
  }
});

Blackprint.allowModuleOrigin('*'); // Allow load from any URL (localhost/https only)

let instance = new Blackprint.Engine();
let remote = new Blackprint.RemoteEngine(instance);
remote.on('module.add', ({ list }) => {
	console.log(`Adding ${list.length} new module, triggered by remote sync`);
});
remote.on('module.added', ({ list, failed }) =>{
	console.log(`${list.length} new module has been added`);

	if(failed.length !== 0)
		console.log(`Failed to add ${failed.length} new module`);
});
remote.on('module.remove', ({ list }) => {
	console.log(`${list.length} module has been removed, triggered by remote sync`);
});
remote.on('disabled', ()=> console.log('Due to some reason, remote control was disabled'));

// Allow import/module sync (return true = allow, false = disable sync)
remote.onImport = v=> console.log("Remote import is allowed") || true;
remote.onModule = v=> console.log("Remote module is allowed") || true;

// "Blackprint.onModuleConflict" need to be replaced if you want to use this to solve conflicting nodes
// Use below if you want to always use the newest module
// Blackprint.onModuleConflict = async map => Object.entries(map).forEach(v => v.useOld = false);

let engineStartup = Date.now();
io.on('connection', client => {
	client.on('relay', data => remote.onSyncIn(data));
	remote.onSyncOut = data => client.emit('relay', data);

	console.log('Remote control: connected');
	client.on('disconnect', () => {
		console.log('Remote control: disconnected');
	});

	setTimeout(() => {
		client.emit('startup-time', engineStartup);
	}, 1000);
});

console.log(`Waiting connection on port: ${port}`);
httpServer.listen(port);

// Blackprint.loadModuleFromURL([
// 	"http://localhost:6789/dist/nodes-console.mjs",
// 	"http://localhost:6789/dist/nodes-data.mjs",
// ]);

// Blackprint.loadModuleFromURL([
// 	"http://localhost:6789/dist/nodes-example.mjs",
// ]);

File: python/example/remote-engine.py
import Blackprint.ModuleLoader
from aiohttp import web
import asyncio
import socketio
import BlackprintRC
import Blackprint
import datetime

import os
ignore_directory = { 'Blender' }
Blackprint.ModuleLoader.add_path_hot_reload(os.path.dirname(__file__) + '/BPNode', lambda path: remote.notifyPuppetNodeListReload({'file': path}), ignore_directory)

instance = Blackprint.Engine()
remote = BlackprintRC.RemoteEngine(instance)

def moduleAdd(ev):
	print(f"Adding {ev.list.length} new module, triggered by remote sync")
remote.on('module.add', moduleAdd)

def moduleAdded(ev):
	print(f"{ev.list.length} new module has been added")

	if(ev.failed.length != 0):
		print(f"Failed to add {ev.failed.length} new module")
remote.on('module.added', moduleAdded)

def moduleRemove(ev):
	print(f"{ev.list.length} module has been removed, triggered by remote sync")
remote.on('module.remove', moduleRemove)

remote.on('disabled', lambda: print('Due to some reason, remote control was disabled'))

# Allow import/module sync (return true = allow, false = disable sync)
remote.onImport = lambda: print("Remote import is allowed") or True
remote.onModule = lambda: print("Remote module is allowed") or True



#  ========= Socket.IO =========

sio = socketio.AsyncServer(cors_allowed_origins='*')
app = web.Application()
sio.attach(app)

engineStartup = int(datetime.datetime.now().timestamp()*1000)
loop = None

@sio.event
async def connect(sid, environ):
	global loop
	if(loop == None): loop = asyncio.get_running_loop()

	remote.onSyncOut = lambda data: asyncio.run_coroutine_threadsafe(sio.emit('relay', data, room=sid), loop)
	await sio.emit('startup-time', engineStartup, room=sid)
	remote.notifyPuppetNodeListReload()
	print('Remote control: connected', sid)

@sio.event
async def relay(sid, data):
	try:
		remote.onSyncIn(data)
	except Exception as e:
		import traceback
		traceback.print_exception(type(e), e, e.__traceback__)

@sio.event
def disconnect(sid):
	print('Remote control: disconnected', sid)

async def index(request):
	return web.Response(text="You need to access this engine from Blackprint Editor", content_type='text/html')

app.router.add_get('/', index)

if __name__ == '__main__':
	web.run_app(app, host="localhost", port=2345)